presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -189,6 +214,80 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, Lookup lookup)
                     true));
         }
 
+        private Optional<DecorrelationResult> rewriteLimitWithRowCountGreaterThanOne(DecorrelationResult childDecorrelationResult, long limit)
+        {
+            PlanNode decorrelatedChildNode = childDecorrelationResult.node;
+
+            // rewrite Limit to RowNumberNode partitioned by constant symbols
+            RowNumberNode rowNumberNode = new RowNumberNode(
+                    idAllocator.getNextId(),
+                    decorrelatedChildNode,
+                    ImmutableList.copyOf(childDecorrelationResult.getConstantSymbols()),
```

Comment:
This is wrong. Think about:
```
corr_x = a
corr_y < b
```
and `LIMIT n`
For each pair of `(corr_x, corr_y)` you want to have `n` rows, but if you just partition by `a` you will flatten such pairs.

I think you have to have condition that all `symbolsToPropagate` are also constants. Please also add test.